src/Elm/Parser/TypeAnnotation.elm

```diff
@@ -13,26 +13,24 @@ import Parser as Core exposing ((|=))
 
 typeAnnotation : Parser State (Node TypeAnnotation)
 typeAnnotation =
-    typeAnnotationNoFnIncludingTypedWithArgumentsLazy
-        |> Combine.andThen
-            (\typeRef ->
-                Layout.optimisticLayoutWith
-                    (\() -> typeRef)
-                    (\() ->
-                        Combine.oneOf
-                            [ arrowRightToTypeAnnotation
-                                |> Combine.map (\ta -> Node.combine TypeAnnotation.FunctionTypeAnnotation typeRef ta)
-                            , Combine.succeed typeRef
-                            ]
-                    )
+    Combine.map (\ta -> \fromTa -> fromTa ta)
```

Comment:
Pretty cool trick!

Reply:
Sadly so far it doesn't seem to show any improvements for me. I'm in the process of converting most andThens to this style and we'll see if this'll work.

Smaller benchmarks I've made seem to prefer the style where each branch actually parses to a union type of what it knows, then leaves the "caller" to attach their info.
This would lead to a lot of temporary types being created (which to reiterate is faster but might be a bit clutter-y)